overlord/devicestate/devicestate.go

```diff
@@ -345,6 +344,21 @@ func remodelTasks(ctx context.Context, st *state.State, current, new *asserts.Mo
 		}
 		tss = append(tss, ts)
 	}
+	// gadget
+	if current.Gadget() == new.Gadget() && current.GadgetTrack() != new.GadgetTrack() {
+		ts, err := snapstateUpdateWithDeviceContext(st, new.Gadget(), &snapstate.RevisionOptions{Channel: new.GadgetTrack()}, userID, snapstate.Flags{NoReRefresh: true}, deviceCtx, fromChange)
```

Comment:
As discussed on IRC, we'll add `needsInstall` pattern from #7715 once, both this one and the other PR land.